agent/sqlib/sqhook/hook.go

```diff
@@ -0,0 +1,285 @@
+// 			// .. function code ...
+//		}
+//
+//
+// Main requirements:
+// - Concurrent access and modification of callbacks.
+// - Reentrant implementation of callbacks with a call context when data needs
+//   to be shared between the prolog and epilog.
+//
+// - Fast call dispatch for callbacks that don't need to be generic, ie.
+//   callbacks that are designed to be attached to specific functions.
+//   Type-assertion instead of `reflect.Call` is therefore used while generic
+//   callbacks that are not tied a specific function will be attached using
+//   `reflect.MakeFunc` in order to match the function signature. The usage
+//   of dynamic calls using `reflect` is indeed much slower and consumes
+//   memory.
+//
+// Design constraints:
+//
+// - There are no compilation-time functions or macros that would have allowed
+//   to provide helpers setting up the hooks in the function definitions.
+// - Access and modification of callbacks need to be atomic.
+// - There are no way to add custom sections to the binary file, which would
+//   have made possible defining the index of hooks at compilation-time (
+//   things that can be easily done with GCC).
+//
+package sqhook
+
+import (
+	"fmt"
+	"reflect"
+	"runtime"
+	"sync/atomic"
+	"unsafe"
+
+	"github.com/sqreen/go-agent/agent/sqlib/sqerrors"
+)
+
+var index = make(map[string]*Hook)
+
+type Hook struct {
+	// The function type where the hook is used.
+	fnType reflect.Type
+	// Pointer to a structure containing the callbacks in order to be able to
+	// atomically modify the pointer.
+	attached *callbacks
+}
+
+type callbacks struct {
+	prolog, epilog Callback
+}
+
+// Callback is a function expecting a Context pointer as first argument,
+// followed by the pointers to the arguments of the hooked function for a
+// prolog, and followed by the pointers to the returned values for an epilog.
+type Callback interface{}
+
+// Context is a call context for the hook. It is shared between the prolog and
+// epilog and is unique for each function call. It allows callbacks to provide
+// reentrant implementations when memory needs to be shared for a given call.
+type Context []interface{}
+
+// MethodReceiver is store in the context when hooking a method.
+type MethodReceiver interface{}
+
+type Error int
+
+// Errors that hooks can return in order to modify the control flow of the
+// function.
+const (
+	_ Error = iota
+	// Abort the execution of the function by returning from it.
+	AbortError
+)
+
+func (e Error) Error() string {
+	return fmt.Sprintf("Error(%d)", e)
+}
+
+// Static assertion that `Error` implements interface `error`
+var _ error = Error(0)
+
+// New returns a hook for function `fn` to be used in the function definition
+// in order to be able to attach callbacks to it. It returns nil if the fn is
+// not a non-nil function or if the symbol name of `fn` cannot be retrieved.
+func New(fn interface{}) *Hook {
+	// Check fn is a non-nil function value.
+	if fn == nil {
+		return nil
```

Comment:
Avoid returning nil. Nillable types are a bad thing in general, better to error/panic as soon as we detect an issue.

Reply:
Why it makes sense here despite the never-make-the-app-fail rule we have:

1. It would be caught at Go initialization time and should therefore be impossible to avoid, unless the program is deployed into production straight away.

2. It should avoid thinking hook work while they don't and therefore avoid not being able to attach a rule relying on a failed hook. Note that failed hooks are not supposed to happen unless there is a bug in the upcoming tool that will insert them.

For these two reasons, panic-ing makes indeed perfect sense.